### ISSUE TITLE
Using rpmalloc in a single thread application

### ISSUE THREAD

**Issue by edubart** [mjansson/rpmalloc#180]
I'm using a rpmalloc in a single thread application, how to proper configure it for that use case to squeeze more performance?
I've used the defines `-DENABLE_THREAD_CACHE=0 -DENABLE_GLOBAL_CACHE=0`, but I notice that still uses some multi-thread specific code. 

I wish there was a way to completely disable multi threading handling via an ifdef,  rpmalloc had the best performance in my single thread use case when compared with others memory allocators (mimalloc, tcmalloc, jemalloc), I think it would could even perform better if there was a way to disable all multi thread handling code.

**Comment by mjansson** [comment]
You most likely want the thread cache to avoid mapping pages with an OS round trip all the time. The global cache however can be disabled.

Other than that it should be pretty much done, the entire idea with the allocator is to allow threads to operate on their own heaps without interfering with other threads, so most of the code is actually "single threaded".

**Comment by edubart** [comment]
But aren't there some overhead in the atomic operations and _Thread_local variables? Also I think it would allow using this library in a C99 compiler that does not support C11 yet.

**Comment by mjansson** [comment]
Yeah, you can try declaring the thread heap as a normal static variable `static heap_t* _memory_thread_heap;` instead and see if it has any measurable impact on your performance.

**Comment by LADSoft** [comment]
yeah I wanted to use rpmalloc in my C++ compiler runtime library (https://github.com/ladsoft/orangec) but immediately ran into a problem:   one of the programs I wanted to compile (the 'omake' program in the distro)  is a multithreaded c++ program on windows (using CreateThread and so forth)  and the separate heaps cause problems for it.   My interpretation is that the 'main' thread passes allocated c++ objects into 'worker' threads and then the worker threads modify them which leads to problems as the main thread data isn't visible to the allocator when a working thread is running...   crashes miserably sigh.    I failed miserably at my attempts to coerce rpmalloc into thinking there was only one thread...   so something like this would have been handy...

**Comment by mjansson** [comment]
The allocator is thread safe and memory blocks can be allocated in one thread and deallocated in another. But tests internal in this repo and the benchmark suite does this.

You could potentially use the first-class heap functions if you want to have direct control of which heap each thread uses.

If you run into problems it's most likely due to errors such as use-after-free or writing/reading out of block boundaries in the code using the allocator.

**Comment by LADSoft** [comment]
it has been a while...   but I think I had to resort to forcing it to use a single heap and then I wrapped all the allocations with critical sections or something similar.    After that it was fine...    but my problem may have been something like allocate on one thread, deallocate on another, which totally wouldn't work with the multiple heaps scheme...    hard to remember...

**Comment by mjansson** [comment]
Well if you ever look at it again, the allocator is definitely capable of handling cross thres allocations/deallocations

**Comment by edubart** [comment]
I made a fork of rpmalloc to be used in single thread applications only, it's my workaround for this issue, I've called it [srpmalloc](https://github.com/edubart/srpmalloc). It's basically a rpmalloc trimmed without atomics and thread APIs, this also makes rpmalloc to work on old C compilers because C11 is not required. 

Maybe in the future rpmalloc could have something like this, a single thread mode ifdef, where all pthread/atomic operations are disabled, I've also find useful to disable huge pages in that fork, to not need to link `advapi32` on Windows and not depend on some platform specific APIs, thus making rpmalloc a little more portable.

I do like to use `rpmalloc` even without multi threading because it seems like to be faster than system's `malloc` in a single thread application.
